File: implementation/scipamato/core/entity/src/test/java/ch/difty/scipamato/core/entity/UserTest.java
package ch.difty.scipamato.core.entity;

import static ch.difty.scipamato.common.entity.ScipamatoEntity.ScipamatoEntityFields.CREATED;
import static ch.difty.scipamato.common.entity.ScipamatoEntity.ScipamatoEntityFields.MODIFIED;
import static ch.difty.scipamato.core.entity.CoreEntity.CoreEntityFields.CREATOR_ID;
import static ch.difty.scipamato.core.entity.CoreEntity.CoreEntityFields.MODIFIER_ID;
import static org.assertj.core.api.Assertions.assertThat;

import java.util.Arrays;
import java.util.Collections;
import java.util.HashSet;
import java.util.Set;

import nl.jqno.equalsverifier.EqualsVerifier;
import nl.jqno.equalsverifier.Warning;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Disabled;
import org.junit.jupiter.api.Test;

import ch.difty.scipamato.core.auth.Role;

class UserTest {

    private static final int     ID         = 1;
    private static final String  USER_NAME  = "username";
    private static final String  FIRST_NAME = "firstname";
    private static final String  LAST_NAME  = "lastname";
    private static final String  EMAIL      = "email";
    private static final String  PASSWORD   = "password";
    private static final boolean ENABLED    = true;

    private final User user = new User(ID, USER_NAME, FIRST_NAME, LAST_NAME, EMAIL, PASSWORD);

    private final Set<Role> roles = new HashSet<>();

    private final Role role1 = Role.ADMIN;
    private final Role role2 = Role.USER;
    private final Role role3 = Role.VIEWER;

    @BeforeEach
    void setUp() {
        roles.addAll(Arrays.asList(role1, role2));
    }

    @Test
    void setAndGet() {
        User u = new User();
        u.setId(ID);
        u.setUserName(USER_NAME);
        u.setFirstName(FIRST_NAME);
        u.setLastName(LAST_NAME);
        u.setEmail(EMAIL);
        u.setPassword(PASSWORD);
        u.setEnabled(ENABLED);
        u.setRoles(roles);

        assertUser(u);
    }

    private void assertUser(User u) {
        assertThat(u.getId()).isEqualTo(1);
        assertThat(u.getUserName()).isEqualTo(USER_NAME);
        assertThat(u.getFirstName()).isEqualTo(FIRST_NAME);
        assertThat(u.getLastName()).isEqualTo(LAST_NAME);
        assertThat(u.getEmail()).isEqualTo(EMAIL);
        assertThat(u.getPassword()).isEqualTo(PASSWORD);
        assertThat(u.isEnabled()).isEqualTo(ENABLED);
        assertThat(u.getRoles()).containsExactlyInAnyOrder(role1, role2);
        assertThat(u.getFullName()).isEqualTo(FIRST_NAME + " " + LAST_NAME);
    }

    @Test
    void constructingByUser() {
        final User u = new User(ID, USER_NAME, FIRST_NAME, LAST_NAME, EMAIL, PASSWORD, ENABLED, roles);
        final User u1 = new User(u);
        assertUser(u1);
    }

    @Test
    void constructingByUser2() {
        final User u = new User(ID, USER_NAME, FIRST_NAME, LAST_NAME, EMAIL, PASSWORD);
        assertThat(u.isEnabled()).isFalse();
        assertThat(u.getRoles())
            .isNotNull()
            .isEmpty();

        u.setEnabled(true);
        u.setRoles(roles);
        assertUser(u);
    }

    @Test
    void displayValue_isEqualToName() {
        final User u = new User(ID, USER_NAME, FIRST_NAME, LAST_NAME, EMAIL, PASSWORD, ENABLED, roles);
        assertThat(u.getDisplayValue()).isEqualTo(USER_NAME);
    }

    @Test
    void reducedConstructor_leavesSomeFieldsDefault() {
        assertThat(user.getId()).isEqualTo(1);
        assertThat(user.getUserName()).isEqualTo(USER_NAME);
        assertThat(user.getFirstName()).isEqualTo(FIRST_NAME);
        assertThat(user.getLastName()).isEqualTo(LAST_NAME);
        assertThat(user.getEmail()).isEqualTo(EMAIL);
        assertThat(user.getPassword()).isEqualTo(PASSWORD);

        assertThat(user.isEnabled()).isFalse();
        assertThat(user.getRoles()).isEmpty();
    }

    @Test
    void settingRoles_reSetsRoles() {
        final User u = new User(ID, USER_NAME, FIRST_NAME, LAST_NAME, EMAIL, PASSWORD, ENABLED, roles);
        assertThat(u.getRoles()).containsExactlyInAnyOrder(role1, role2);

        u.setRoles(Set.of(role3, role1));

        assertThat(u.getRoles()).containsExactlyInAnyOrder(role3, role1);
    }

    @Test
    void settingRole_withNullList_clearsList() {
        final User u = new User(ID, USER_NAME, FIRST_NAME, LAST_NAME, EMAIL, PASSWORD, ENABLED, roles);
        u.setRoles(null);
        assertThat(u.getRoles()).isEmpty();
    }

    @Test
    void settingRole_withBlankList_clearsList() {
        final User u = new User(ID, USER_NAME, FIRST_NAME, LAST_NAME, EMAIL, PASSWORD, ENABLED, roles);
        u.setRoles(Collections.emptySet());
        assertThat(u.getRoles()).isEmpty();
    }

    @Test
    void removingAssignedRole_removesIt() {
        final User u = new User(ID, USER_NAME, FIRST_NAME, LAST_NAME, EMAIL, PASSWORD, ENABLED, roles);
        u.removeRole(role1);
        assertThat(u.getRoles()).containsExactly(role2);
    }

    @Test
    void removingUnassignedRole_doesNothing() {
        final User u = new User(ID, USER_NAME, FIRST_NAME, LAST_NAME, EMAIL, PASSWORD, ENABLED, roles);
        u.removeRole(role3);
        assertThat(u.getRoles()).containsExactlyInAnyOrder(role1, role2);
    }

    @Test
    void addingNullRole_addsNothing() {
        user.addRole(role1);
        assertThat(user.getRoles()).containsExactly(role1);
    }

    @Test
    void equals() {
        EqualsVerifier
            .forClass(User.class)
            .withRedefinedSuperclass()
            .usingGetClass()
            .withIgnoredFields(CREATED.getFieldName(), CREATOR_ID.getFieldName(), MODIFIED.getFieldName(),
                MODIFIER_ID.getFieldName())
            .suppress(Warning.STRICT_INHERITANCE, Warning.NONFINAL_FIELDS)
            .verify();
    }

    @Test
    @Disabled("TODO")
    void testingToString() {
        assertThat(user.toString()).isEqualTo(
            "User[userName=username,firstName=firstname,lastName=lastname,email=email,password=password,enabled=false,roles=[],id=1,createdBy=<null>,lastModifiedBy=<null>,created=<null>,lastModified=<null>,version=0]");
    }

    @Test
    void assertEnumFields() {
        assertThat(User.UserFields.values())
            .extracting("name")
            .containsExactly("userName", "firstName", "lastName", "email", "password", "enabled", "roles");
    }

}

File: implementation/scipamato/core/entity/src/test/kotlin/ch/difty/scipamato/core/entity/UserTest.kt
package ch.difty.scipamato.core.entity

import ch.difty.scipamato.common.entity.ScipamatoEntity.ScipamatoEntityFields.CREATED
import ch.difty.scipamato.common.entity.ScipamatoEntity.ScipamatoEntityFields.MODIFIED
import ch.difty.scipamato.core.auth.Role
import ch.difty.scipamato.core.entity.CoreEntity.CoreEntityFields.CREATOR_ID
import ch.difty.scipamato.core.entity.CoreEntity.CoreEntityFields.MODIFIER_ID
import nl.jqno.equalsverifier.EqualsVerifier
import nl.jqno.equalsverifier.Warning
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Disabled
import org.junit.jupiter.api.Test
import java.util.*

internal class UserTest {

    private val user = User(ID, USER_NAME, FIRST_NAME, LAST_NAME, EMAIL, PASSWORD)

    private val roles = HashSet<Role>()

    private val role1 = Role.ADMIN
    private val role2 = Role.USER
    private val role3 = Role.VIEWER

    @BeforeEach
    fun setUp() {
        roles.addAll(Arrays.asList(role1, role2))
    }

    @Test
    fun setAndGet() {
        val u = User()
        u.id = ID
        u.userName = USER_NAME
        u.firstName = FIRST_NAME
        u.lastName = LAST_NAME
        u.email = EMAIL
        u.password = PASSWORD
        u.isEnabled = ENABLED
        u.roles = roles

        assertUser(u)
    }

    private fun assertUser(u: User) {
        assertThat(u.id).isEqualTo(1)
        assertThat(u.userName).isEqualTo(USER_NAME)
        assertThat(u.firstName).isEqualTo(FIRST_NAME)
        assertThat(u.lastName).isEqualTo(LAST_NAME)
        assertThat(u.email).isEqualTo(EMAIL)
        assertThat(u.password).isEqualTo(PASSWORD)
        assertThat(u.isEnabled).isEqualTo(ENABLED)
        assertThat(u.roles).containsExactlyInAnyOrder(role1, role2)
        assertThat(u.fullName).isEqualTo("$FIRST_NAME $LAST_NAME")
    }

    @Test
    fun constructingByUser() {
        val u = User(ID, USER_NAME, FIRST_NAME, LAST_NAME, EMAIL, PASSWORD, ENABLED, roles)
        val u1 = User(u)
        assertUser(u1)
    }

    @Test
    fun constructingByUser2() {
        val u = User(ID, USER_NAME, FIRST_NAME, LAST_NAME, EMAIL, PASSWORD)
        assertThat(u.isEnabled).isFalse()
        assertThat(u.roles)
                .isNotNull
                .isEmpty()

        u.isEnabled = true
        u.roles = roles
        assertUser(u)
    }

    @Test
    fun displayValue_isEqualToName() {
        val u = User(ID, USER_NAME, FIRST_NAME, LAST_NAME, EMAIL, PASSWORD, ENABLED, roles)
        assertThat(u.displayValue).isEqualTo(USER_NAME)
    }

    @Test
    fun reducedConstructor_leavesSomeFieldsDefault() {
        assertThat(user.id).isEqualTo(1)
        assertThat(user.userName).isEqualTo(USER_NAME)
        assertThat(user.firstName).isEqualTo(FIRST_NAME)
        assertThat(user.lastName).isEqualTo(LAST_NAME)
        assertThat(user.email).isEqualTo(EMAIL)
        assertThat(user.password).isEqualTo(PASSWORD)

        assertThat(user.isEnabled).isFalse()
        assertThat(user.roles).isEmpty()
    }

    @Test
    fun settingRoles_reSetsRoles() {
        val u = User(ID, USER_NAME, FIRST_NAME, LAST_NAME, EMAIL, PASSWORD, ENABLED, roles)
        assertThat(u.roles).containsExactlyInAnyOrder(role1, role2)

        u.roles = setOf(role3, role1)

        assertThat(u.roles).containsExactlyInAnyOrder(role3, role1)
    }

    @Test
    fun settingRole_withNullList_clearsList() {
        val u = User(ID, USER_NAME, FIRST_NAME, LAST_NAME, EMAIL, PASSWORD, ENABLED, roles)
        u.roles = null
        assertEquals(0, u.roles.size)
    }

    @Test
    fun settingRole_withBlankList_clearsList() {
        val u = User(ID, USER_NAME, FIRST_NAME, LAST_NAME, EMAIL, PASSWORD, ENABLED, roles)
        u.roles = emptySet()
        assertThat(u.roles).isEmpty()
    }

    @Test
    fun removingAssignedRole_removesIt() {
        val u = User(ID, USER_NAME, FIRST_NAME, LAST_NAME, EMAIL, PASSWORD, ENABLED, roles)
        u.removeRole(role1)
        assertThat(u.roles).containsExactly(role2)
    }

    @Test
    fun removingUnassignedRole_doesNothing() {
        val u = User(ID, USER_NAME, FIRST_NAME, LAST_NAME, EMAIL, PASSWORD, ENABLED, roles)
        u.removeRole(role3)
        assertThat(u.roles).containsExactlyInAnyOrder(role1, role2)
    }

    @Test
    fun addingNullRole_addsNothing() {
        user.addRole(role1)
        assertThat(user.roles).containsExactly(role1)
    }

    @Test
    fun equals() {
        EqualsVerifier
                .forClass(User::class.java)
                .withRedefinedSuperclass()
                .usingGetClass()
                .withIgnoredFields(CREATED.fieldName, CREATOR_ID.fieldName, MODIFIED.fieldName, MODIFIER_ID.fieldName)
                .suppress(Warning.STRICT_INHERITANCE, Warning.NONFINAL_FIELDS)
                .verify()
    }

    @Test
    @Disabled("TODO")
    fun testingToString() {
        assertThat(user.toString()).isEqualTo(
                "User[userName=username,firstName=firstname,lastName=lastname,email=email,password=password,enabled=false,roles=[],id=1,createdBy=<null>,lastModifiedBy=<null>,created=<null>,lastModified=<null>,version=0]")
    }

    @Test
    fun assertEnumFields() {
        assertThat(User.UserFields.values()).extracting("name")
                .containsExactly("userName", "firstName", "lastName", "email", "password", "enabled", "roles")
    }

    companion object {
        private val ID = 1
        private val USER_NAME = "username"
        private val FIRST_NAME = "firstname"
        private val LAST_NAME = "lastname"
        private val EMAIL = "email"
        private val PASSWORD = "password"
        private val ENABLED = true
    }

}
